File: Data_Structures/Tree/Binary_Tree_Traversal/Binary_Tree_Traversal_Iterative.cpp
#include <cstdio>
#include <stack>

typedef struct node node;

struct node {
  int data;
  node *left, *right;
  node(int data) {
    this->data = data;
    left = right = nullptr;
  }
};

void InOrder(node *root) {
  std::stack<node*> S;
  do {
    while (root != nullptr) {
      S.push(root);
      root = root->left;
    }

    if (!S.empty()) {
      root = S.top();
      S.pop();

      printf("%d-", root->data);
      root = root->right;
    }
  } while (!S.empty() || root != nullptr);
  printf("\n");
}

void PreOrder(node *root) {
  std::stack<node*> S;
  do {
    while (root != nullptr) {
      S.push(root);
      printf("%d-", root->data);
      root = root->left;
    }

    if (!S.empty()) {
      root = S.top();
      S.pop();
      root = root->right;
    }
  } while (!S.empty() || root != nullptr);
  printf("\n");
}

void PostOrder(node *root) {
  std::stack<node*> S;
  node *prev = nullptr;
  do {
    while (root != nullptr) {
      S.push(root);
      root = root->left;
    }
    while (root == nullptr && !S.empty()) {
      root = S.top();
      if (root->right == nullptr || root->right == prev) {
        printf("%d-", root->data);
        S.pop();
        prev = root;
        root = nullptr;
      } else
        root = root->right;
    }
  } while (!S.empty());
  printf("\n");
}

int main() {
  node* n= new node(5);
  n->right = new node(10);
  n->left = new node(12);
  n->right->left = new node(4);
  n->right->right = new node(50);
  n->left->right = new node(78);

  InOrder(n);
  PreOrder(n);
  PostOrder(n);

  return 0;
}

File: Data_Structures/Tree/Binary_Tree_Traversal/Binary_Tree_Traversal_Iterative.js

class Node {
  constructor({value, left, right}) {
    this.value = value;
    this.left = left;
    this.right = right;
  }
}

const isNotEmpty = stack => stack.length > 0;

class BinaryTree {
  constructor(root) {
    this.root = root;
  }

  visit(f) {
  }

  preOrder() {
    const result = [];
    const stack = [];
    let node = this.root;
    do {
      while (node !== undefined) {
        result.push(node.value);
        stack.push(node);
        node = node.left;
      }

      if (isNotEmpty(stack)) {
        node = stack.pop();
        node = node.right;
      }
    } while (isNotEmpty(stack) || node !== undefined);
    return result;
  }

  postOrder() {
    const result = [];
    const stack = [];
    let node = this.root;
    let prev = undefined;
    do {
      while (node !== undefined) {
        stack.unshift(node);
        node = node.left;
      }

      while (node === undefined && isNotEmpty(stack)) {
        node = stack[0];
        if (node.right === undefined || node.right == prev) {
          result.push(node.value);
          stack.shift();
          prev = node;
          node = undefined;
        } else {
          node = node.right;
        }
      }
    } while (isNotEmpty(stack));
    return result;
  }

  inOrder() {
    const result = [];
    const stack = [];
    let node = this.root;
    do {
      while (node !== undefined) {
        stack.push(node);
        node = node.left;
      }

      if (isNotEmpty(stack)) {
        node = stack.pop();
        result.push(node.value);
        node = node.right;
      }
    } while (isNotEmpty(stack) || node !== undefined);
    return result;
  }
}

const buildTree = () => {
  const node7 = new Node({value: 7});
  const node6 = new Node({value: 6});
  const node5 = new Node({value: 5});
  const node4 = new Node({value: 4, right: node7});
  const node2 = new Node({value: 2, left: node4, right: node5});
  const node3 = new Node({value: 3, right: node6});
  const node1 = new Node({value: 1, left: node2, right: node3});
  return node1;
};

const main = () => {
  const tree = new BinaryTree(buildTree());
  console.log("tree =",tree);
  console.log("pre-order", tree.preOrder());
  console.log("post-order", tree.postOrder());
  console.log("in-order", tree.inOrder());
};

main();
